libssh-rs/src/lib.rs

```diff
@@ -274,6 +282,55 @@ impl Session {
         }
     }
 
+    unsafe extern "C" fn bridge_channel_open_request_auth_agent_callback(
+        session: sys::ssh_session,
+        userdata: *mut ::std::os::raw::c_void,
+    ) -> sys::ssh_channel {
+        let result = std::panic::catch_unwind(|| -> SshResult<sys::ssh_channel> {
+            let sess: &mut SessionHolder = &mut *(userdata as *mut SessionHolder);
+            assert!(
+                std::ptr::eq(session, sess.sess),
+                "invalid callback invocation: session mismatch"
+            );
+            let cb = sess
+                .channel_open_request_auth_agent_callback
+                .as_mut()
+                .unwrap();
+            let chan = unsafe { sys::ssh_channel_new(session) };
+            if chan.is_null() {
+                return Err(sess
+                    .last_error()
+                    .unwrap_or_else(|| Error::fatal("ssh_channel_new failed")));
+            }
+            match cb(Channel::new(&sess.outer.upgrade().unwrap(), chan)) {
```

Comment:
In general, I like to avoid adding `unwrap`s when the calling scope can return a Result.
Can you re-phrase this using the `?` operator?

Reply:
Done.

However an upgrade failure really should not happen here, it means a channel outlived the corresponding session, code looks like this now:

```rs
.ok_or_else(|| Error::fatal("impossible: channel callback outlived session"))?
```

Do you still prefer returning an Error, or should I use `.expect()`?

Reply:
Code removed in latest version.